packages/server/lib/util/resourceTypeAndCredentialManager.ts

```diff
@@ -0,0 +1,88 @@
+import md5 from 'md5'
+import Debug from 'debug'
+import type { RequestCredentialLevel, RequestResourceType } from '@packages/proxy'
+
+type AppliedCredentialByUrlAndResourceMap = Map<string, Array<{
+  resourceType: RequestResourceType
+  credentialStatus: RequestCredentialLevel
+}>>
+
+const debug = Debug('cypress:server:util:resource-type-and-credential')
+
+const hashUrl = (url: string): string => {
+  return md5(decodeURIComponent(url))
+}
+
+// leverage a singleton Map throughout the server to prevent clashes with this context bindings
+const _appliedCredentialByUrlAndResourceMap: AppliedCredentialByUrlAndResourceMap = new Map()
+
+class ResourceTypeAndCredentialManagerClass {
```

Comment:
```suggestion
class ResourceTypeAndCredentialManager {
```

Just a nit, but otherwise it seems redundant.

Reply:
added some comments in #24060. I'm going to keep it as `ResouceTypeAndCredentialManagerClass` but export the type `ResourceTypeAndCredentialManager` so others do not try to import the class and instantiate it.